File: leetcode/test10.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;


class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.size();
        int n = p.size();

        auto matches = [&](int i, int j) {
            if (i == 0) {
                return false;
            }
            if (p[j - 1] == '.') {
                return true;
            }
            return s[i - 1] == p[j - 1];
        };
        // f[0][0]表示空字符串，f[i][j]表示s的第i个字符于p的第j个字符是否匹配，i/j从1开始，所以在s/p中需要判断s[i-1]与p[j-1]的值

        vector<vector<int>> f(m + 1, vector<int>(n + 1));
        f[0][0] = true;
        for (int i = 0; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                if (p[j - 1] == '*') {
                    f[i][j] |= f[i-1][j - 1];
                    if (matches(i, j - 1)) {
                        f[i][j] |= f[i - 1][j];
                    }
                }
                else {
                    if (matches(i, j)) {
                        f[i][j] |= f[i - 1][j - 1];
                    }
                }
            }
        }
        return f[m][n];
    }
};


int main() {
    std::cout << "Hello, World!" << std::endl;
    Solution s;
    std::cout << s.isMatch("abc", "ab*");
    return 0;
}

File: leetcode_py/test10.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        A = [[False for j in range(0, len(p)+1)] for i in range(0, len(s)+1)]
        A[0][0] = True
        def match(i, j):
            if i == 0:
                return False
            if p[j-1] == '.':
                return True
            return s[i-1] == p[j-1]
        for i in range(0, len(s)+1):
            for j in range(1, len(p)+1):
                if(p[j-1]!='*'):
                    if match(i, j):
                        A[i][j] = A[i-1][j-1]
                    else:
                        A[i][j] = False
                else:
                    if match(i, j-1):
                        A[i][j] = A[i-1][j] or A[i][j-2]
                    else:
                        A[i][j] = A[i][j-2]
        return A[-1][-1]


if __name__ == '__main__':
    s = Solution()
    print(s.isMatch('a', '.*'))